apps/rabbitmq_prelaunch/src/rabbit_prelaunch_errors.erl

```diff
@@ -0,0 +1,109 @@
+                      Unexpected, Node}}) ->
+    rabbit_misc:format(
+      "Erlang distribution running with another node name (~s) "
+      "than the configured one (~s)",
+      [Unexpected, Node]);
+format_error({bad_config_entry_decoder, missing_passphrase}) ->
+    rabbit_misc:format(
+      "Missing passphrase or missing passphrase read method in "
+      "`config_entry_decoder`");
+format_error({config_decryption_error, {key, Key}, _Msg}) ->
+    rabbit_misc:format(
+      "Error while decrypting key '~p'. Please check encrypted value, "
+      "passphrase, and encryption configuration~n",
+      [Key]);
+format_error({error, {timeout_waiting_for_tables, AllNodes, _}}) ->
+    Suffix =
+    "~nBACKGROUND~n==========~n~n"
+    "This cluster node was shut down while other nodes were still running.~n"
+    "To avoid losing data, you should start the other nodes first, then~n"
+    "start this one. To force this node to start, first invoke~n"
+    "\"rabbitmqctl force_boot\". If you do so, any changes made on other~n"
+    "cluster nodes after this one was shut down may be lost.",
+    {Message, Nodes} =
+    case AllNodes -- [node()] of
+        [] -> {rabbit_misc:format(
+                 "Timeout contacting cluster nodes. Since RabbitMQ was"
+                 " shut down forcefully~nit cannot determine which nodes"
+                 " are timing out.~n" ++ Suffix, []),
+               []};
+        Ns -> {rabbit_misc:format(
+                 "Timeout contacting cluster nodes: ~p.~n" ++ Suffix,
+                 [Ns]),
+               Ns}
+    end,
+    Message ++ "\n" ++ rabbit_nodes_common:diagnostics(Nodes);
+format_error({error, {cannot_log_to_file, unknown, Reason}}) ->
+    rabbit_misc:format(
+      "failed to initialised logger: ~p~n",
+      [Reason]);
+format_error({error, {cannot_log_to_file, LogFile,
+                      {cannot_create_parent_dirs, _, Reason}}}) ->
+    rabbit_misc:format(
+      "failed to create parent directory for log file at '~s', reason: ~s~n",
+      [LogFile, file:format_error(Reason)]);
+format_error({error, {cannot_log_to_file, LogFile, Reason}}) ->
+    rabbit_misc:format(
+      "failed to open log file at '~s', reason: ~s",
+      [LogFile, file:format_error(Reason)]);
+format_error(Error) ->
+    rabbit_misc:format("Error during startup: ~p", [Error]).
+
+log_exception(Class, Exception, Stacktrace) ->
+    Message = format_exception(Class, Exception, Stacktrace),
+    log_message(Message).
+
+format_exception(Class, Exception, Stacktrace) ->
+    rabbit_misc:format(
+      "Exception during startup:~n~s",
+      [lager:pr_stacktrace(Stacktrace, {Class, Exception})]).
+
+log_message(Message) ->
+    Lines = string:split(
+              ?BOOT_FAILED_HEADER ++
+              Message ++
+              ?BOOT_FAILED_FOOTER,
+              [$\n],
+              all),
+    [rabbit_log_prelaunch:error("~s", [Line]) || Line <- Lines],
+    [io:format(standard_error, "~s~n", [Line]) || Line <- Lines],
```

Comment:
This generates a duplicated `BOOT FAILED` log message, is it needed?

```
RABBITMQ_NODE_PORT=5673 RABBITMQ_NODE_NAME=rabbit@mars RABBITMQ_DBG=mod=narabbit_channel:info_all,rabbit_channel:info_local TMQ_LOGS="/tmp/kk/rabbit.log" RABBITMQ_FEATURE_FLAGS_FILE="/Users/dparracorbacho/test/ff" RABBITMQ_CONFIG_FILE="/Users/dparracorbacho/test/rabbit" ./sbin/rabbitmq-server
13:10:04.744 [error]
13:10:04.750 [error] BOOT FAILED
13:10:04.750 [error] ===========
13:10:04.750 [error] ERROR: node with name "rabbit" already running on "mars"
13:10:04.750 [error]

BOOT FAILED
===========
ERROR: node with name "rabbit" already running on "mars"

13:10:05.751 [error] Supervisor rabbit_prelaunch_sup had child prelaunch started with rabbit_prelaunch:run_prelaunch_first_phase() at undefined exit with reason {duplicate_node_name,"rabbit","mars"} in context start_error
13:10:05.751 [error] CRASH REPORT Process <0.145.0> with 0 neighbours exited with reason: {{shutdown,{failed_to_start_child,prelaunch,{duplicate_node_name,"rabbit","mars"}}},{rabbit_prelaunch_app,start,[normal,[]]}} in application_master:init/4 line 138
{"Kernel pid terminated",application_controller,"{application_start_failure,rabbitmq_prelaunch,{{shutdown,{failed_to_start_child,prelaunch,{duplicate_node_name,\"rabbit\",\"mars\"}}},{rabbit_prelaunch_app,start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,rabbitmq_prelaunch,{{shutdown,{failed_to_start_child,prelaunch,{duplicate_node_name,"rabbit","mars"}}},{rabbit_prelaunch_app,

Crash dump is being written to: erl_crash.dump...done
```

Reply:
That's because this error happens before logging is fully configured. Any boot error occurring after logging is ready will display one message on `stderr` and one in the log file.

Before logging is configured, yes both messages end up in a terminal, but not on the same file descriptor: the first one is written to `stdout`, the second one is written to `stderr`.

The logging function being generic, regardless of the loggin configuration status, yes, we get the message twice if the output was not redirected. If you try with TCP port 5672 (which is open after logging configuration), you'll get one message on `stderr`, one in the log file.

Do you think it's an issue?

Reply:
No, that's a good enough reason for it. It's fine if sometimes we have duplicates.

Reply:
Thanks for the explanation @dumbbell